File: packages/ssh-native/android/src/main/java/dev/happier/ssh/HappierSshNativeModule.kt
package dev.happier.ssh

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class HappierSshNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HappierSshNative")

    Events("hostKeyPrompt", "authPrompt", "progress")

    Function("getAvailability") {
      return@Function HappierSshNativeBridge.availability()
    }

    AsyncFunction("exec").SuspendBody { request: Map<String, Any?> ->
      withContext(Dispatchers.IO) {
        HappierSshNativeBridge.exec(this@HappierSshNativeModule, request)
      }
    }

    AsyncFunction("respondToHostKeyPrompt") { promptId: String, response: Map<String, Any?> ->
      return@AsyncFunction HappierSshNativeBridge.respondToHostKeyPrompt(promptId, response)
    }

    AsyncFunction("respondToAuthPrompt") { promptId: String, response: Map<String, Any?> ->
      return@AsyncFunction HappierSshNativeBridge.respondToAuthPrompt(promptId, response)
    }

    AsyncFunction("cancelRequest") { requestId: String ->
      return@AsyncFunction HappierSshNativeBridge.cancelRequest(requestId)
    }

    AsyncFunction("startLoopbackTunnel").SuspendBody { request: Map<String, Any?> ->
      withContext(Dispatchers.IO) {
        HappierSshNativeBridge.startLoopbackTunnel(this@HappierSshNativeModule, request)
      }
    }

    AsyncFunction("stopLoopbackTunnel") { nativeTunnelId: String ->
      return@AsyncFunction HappierSshNativeBridge.stopLoopbackTunnel(nativeTunnelId)
    }
  }
}

File: packages/ssh-native/ios/HappierSshNativeModule.swift
import ExpoModulesCore
import Foundation

public final class HappierSshNativeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HappierSshNative")

    Events("hostKeyPrompt", "authPrompt", "progress")

    Function("getAvailability") { () -> [String: Any] in
      HappierSshNativeBridge.availability()
    }

    AsyncFunction("exec") { (_ request: [String: Any]) async throws -> [String: Any] in
      try await HappierSshNativeBridge.execAsync(module: self, request: request)
    }

    AsyncFunction("respondToHostKeyPrompt") { (_ promptId: String, _ response: [String: Any]) -> Void in
      HappierSshNativeBridge.respondToHostKeyPrompt(promptId: promptId, response: response)
    }

    AsyncFunction("respondToAuthPrompt") { (_ promptId: String, _ response: [String: Any]) -> Void in
      HappierSshNativeBridge.respondToAuthPrompt(promptId: promptId, response: response)
    }

    AsyncFunction("cancelRequest") { (_ requestId: String) -> Void in
      HappierSshNativeBridge.cancelRequest(requestId: requestId)
    }

    AsyncFunction("startLoopbackTunnel") { (_ request: [String: Any]) async throws -> [String: Any] in
      try await HappierSshNativeBridge.startLoopbackTunnelAsync(module: self, request: request)
    }

    AsyncFunction("stopLoopbackTunnel") { (_ nativeTunnelId: String) -> Void in
      HappierSshNativeBridge.stopLoopbackTunnel(nativeTunnelId: nativeTunnelId)
    }
  }
}
